clipper_admin/clipper_manager.py

```diff
@@ -628,13 +630,136 @@ def _save_python_function(self, name, predict_function):
+                    ignore_list.append(get_module_name(line))
+
+        if (os.path.isfile(pip_deps_fname)):
+            with open(pip_deps_abs_path, 'r') as f:
+                for line in f:
+                    ignore_list.append(get_module_name(line))
+
+        with open(python_container_conda_deps_fname, 'r') as f:
+            for line in f:
+                ignore_list.append(line.strip())
+
+        return ignore_list
+
+    def _export_local_modules(self, func_modules, serialization_dir,
+                              conda_deps_fname, pip_deps_fname,
+                              local_modules_folder_name):
+        """
+        Supplies modules in `func_modules` not already accounted for by Anaconda or pip to container.
+        Copies those modules to {`serialization_dir`}/{`local_module_deps_fname`}.
+
+        Parameters
+        ----------
+        func_modules : str
+            The modules to consider exporting
+        serialization_dir : str
+            The path to the diretory containing the conda, pip, and output local modules dependency information
+        conda_deps_fname : str
+            The name of the conda dependency file
+        pip_deps_fname : str
+            The name of the pip dependency file
+        local_modules_folder_name : str
+            The name of the folder to copy modules to
+        """
+        ignore_list = self._get_already_exported_modules(
+            serialization_dir, conda_deps_fname, pip_deps_fname)
+        paths_to_copy = set()
+
+        mda = ModuleDependencyAnalyzer()
+        for module_name in ignore_list:
+            mda.ignore(module_name)
+        for module in func_modules:
+            module_name = module.__name__
+            if module_name not in ignore_list and 'clipper_admin' not in module_name:
+                mda.add(module_name)
+
+                if '.' in module_name:
+                    module_name_split = module_name.split()
+                    try:
+                        file_location = module.__file__
+                        package_nested_layer = len(module_name.split('.')) - 1
+                        folder_indices = [
+                            i for i, x in enumerate(file_location) if x == "/"
+                        ]
+                        cutoff_index = folder_indices[-package_nested_layer]
+                        package_path = file_location[:cutoff_index]
+                        paths_to_copy.add(package_path)
+                    except:
+                        print(
+                            "Could not identify the location of the root package {package} for module {module}. Will not supply it to the container.".
+                            format(
+                                package=module_name_split[0],
+                                module=module_name))
+            else:
+                mda.ignore(module_name)
+        module_paths = mda.get_and_clear_paths()
+
+        def path_already_captured(path):
```

Comment:
Let's say your predict function makes use of some package-based imported module: `import my_package.my_module`. Right now, we add the path to `.../my_package/` to `paths_to_copy` and plan on recursively copying the whole folder to the container later. The location of my_package on the container would be: `/model/modules/my_package`. These will be accessible for import because `/model/modules` will be added to container's sys.path in the container code.

We also add that module name, `my_package.my_module`, to our Module Dependency Analyzer to find all the files that the module is dependent on; it's possible that some don't exist in the folder `.../my_package/` and we still want to copy them to the container. For each file with path `.../<file_name>`, its copy location on the container will be `/model/modules/<file_name>`, and these will be accessible to directly import because `/model/modules` will be added to container's sys.path in the container code.

In order to get the paths to all these files, we run `module_paths = mda.get_and_clear_paths()`, `module_paths` will certainly have files contained in `my_package/...`. To avoid copying multiple versions of these files, we check here if each is already captured (using `path_already_captured`), and don't add it to the set of lists we'll copy if it is.

This seems good in most cases because it prevents us from doing a bunch of unnecessary copies. However, there's a case (seems kind of unlikely) where it will cause an ImportError. Let's say that your prediction function makes use of two modules: one imported via package (`import my_package.my_module`), and one imported directly by adding it's enclosing folder to your path and running (`import my_module_2`). Let's say that `my_module_2` resides under `my_package/`.

Now, we'll run into an ImportError in the container because it won't be able to import the module with name `my_module_2`; it expects the module file to be in it's path (probably at `/model/modules/my_module_2.py`, but we prevented it from getting copied there because we were already going to copy it to `/model/modules/my_package/my_module_2.py`.

I'm not sure how much we care about this case. If we don't think it's costly to have a bunch of unnecessary file copies, then I think we should remove the code responsible for avoiding them (lines 736-744 here). Do you have any thoughts?